### ISSUE TITLE
Correct Use Of "Evolving" Context

### ISSUE THREAD

**Issue by dsbw** [cljfx/cljfx#140]
I apologize for the poor title, but I wasn't sure how to describe this: I have a button bar at the top of my app, with an `Open` button. This button needs to be disabled until the user selects an item. The user selects the item and that causes the context to populate with detail under the `selected-item` key. The `Open` button checks the context for that `selected-item` key, and knows to enable itself based on that value being populated.

Problem is, this is done in one handler. So the incoming context doesn't know it's got a selected item: It's both populating the item _and_ calling for the buttons to update. I resolved this by creating an interim context with the selected item and passing _that_ to the `get-buttons` routine which checks for it.

It works but I'm not sure if that's the correct handling or not. I thought about firing a second event to update menus, etc., after updating the base state but that seemed too complicated.

```
(defmethod event-handler :select-item [{:keys [fx/event fx/context] :as e}]
  (let [ino (java.lang.Integer. (event :itemNumber))
        ctx1 (fx/swap-context context assoc-in [:selected-item] (detail item))
        bb (or (get-buttons ctx1 :item-window) (:button-bar (:cljfx.context/m ctx1)))]
    {:context (fx/swap-context ctx1 assoc-in [:button-bar] bb)}))
```

**Comment by dsbw** [comment]
> Hi! I'm not quite following what's going on in the code (what does get-buttons do?),

The button-bar I mentioned acts as a giant, very visible menu, so it changes as the context changes. `get-buttons` returns the correct buttons for the current view and state. Each view registers the buttons it needs:

```
(register-view :ui ui
               [["E_xit" :exit]
                ["_Setup" :navigate {:view :setup}]
                ["_About" :navigate {:view :about}]])
```
I have a function that takes `["_Setup" :navigate {:view :setup}]` and turns it into a button labeled "Setup" that when you click it fires the `:navigate` handler. The `:navigate` handler then changes the `:view` and the `:button-bar`:
```
(defmethod event-handler :navigate [{:keys [fx/context view mode]}]
 {:context (fx/swap-context context #(-> %
                                          (assoc-in [:button-bar] (or (get-buttons context view) (:button-bar (:cljfx.context/m context))))
                                          (assoc-in [:view] (get-view view))))})
```
And this works fine when I'm swapping views around. (Although looking at it now, maybe I should just have the button-bar dependent on the `:view` key so that it updates itself when `:view` is updated. Is that the root of my problem?) But sometimes I need context to update the menu, so I pass in a function to `register-view` instead of just a straight button vector:
```
(register-view :service-writer service-writer
               (fn [context]
                 (let [selected (fx/sub-val context :selected-item)]
                   (vector ["Back to _Main Menu" :navigate {:view :ui}]
                      ["_Select" :navigate {:view :select-item} {:disable (fx/sub-ctx context (fn [x] (not (boolean selected))))}]
                      ["_New" :navigate {:view :new-item :mode :new-item}]
. . .
```
So the "Select" button is disabled as long as the :selected-item key is null. The problem being that when this value is created `[selected (fx/sub-val context :selected-item)]` the context hasn't yet been updated to have a value for `:selected-item`.

> but it seems to me you are accessing JavaFX button instances and mutating them yourself by switching them to enabled/disabled state? 

I'm not, as you can see! Though I would certainly be amenable to the idea that I'm doing things the hard way. (I didn't even know you could GET the JavaFX instances of things.)

**Comment by dsbw** [comment]
Resolved with question #146 .
